`remove`: output file/dir by path (#2357)
iterative/dvc.org: content/docs/command-reference/remove.md

usage: dvc remove [-h] [-q | -v] [--outs] targets [targets ...]

positional arguments:
  targets        Tracked files/directories, stage names (found in
                 dvc.yaml), or .dvc files to remove.
```

## Description

> Safely removes tracked files/directories, stage names (found in `dvc.yaml`), or
> `.dvc` files. This includes deleting the corresponding `.gitignore` entries.

Reviewer: Sorry for the delay. Let's do this here:
```suggestion
Safely removes tracked data (by file name, stage name, or `.dvc` file path).
This includes deleting the corresponding `.gitignore` entries.
```

> `dvc remove` doesn't remove files from the DVC <abbr>cache</abbr> or
> [remote storage](/doc/command-reference/remote). Use `dvc gc` for that.

> 

Reviewer 2: I think it would be great to have an example (maybe just extending the one below) that shows how to remove the data from the cache, although I don't know that current gc functionality is sufficient to really be useful in this scenario.

It takes one or more stage names (see `-n` option of `dvc run`), `.dvc` file
names or tracked files/directories as `targets`.

If there are no stages left in `dvc.yaml` after the removal, then both
`dvc.yaml` and `dvc.lock` are deleted. `.gitignore` is also deleted if there are
no more entries left in it.

Note that, when using stage name as target, the actual <abbr>output</abbr> files
or directories of the stage (`outs` field) are not removed by this command,
unless the `--outs` option is used which will remove **all** of them.

> Alternatively, you can the names of individual <abbr>output</abbr> files or

Reviewer 3: but in this case we don't do anything to the .dvc or dvc.yaml? 

this mixed semantics confuses me a bit to be honest

with some targets we touch dvc.yaml, but not data ... with some targets we touch data, but not DVC files 

also, `--outs` that is not clear how it should be have when file/directory path is provided

It feels to me that even if we merge this (and the DVC core one), we'll need to get back to the whiteboard pretty soon with this `dvc remove` that we are long overdue cc @efiop @dberenbaum

Reviewer 3: Overall, not sure if this has enough value to do this now at least - it seems it's pretty much a replacement to `rm -rf` (+git command to remove .gitignore), unless I'm missing something? @daavoo what is your use case that you had in mind?

sorry, that we didn't get back to you initially in the DVC core ticket (or may be I missed the discussion).

Author: > Overall, not sure if this has enough value to do this now at least - it seems it's pretty much a replacement to `rm -rf` (+git command to remove .gitignore), unless I'm missing something? @daavoo what is your use case that you had in mind?
> 
> sorry, that we didn't get back to you initially in the DVC core ticket (or may be I missed the discussion).

I agree with the problem of mixed semantics @shcheklein . Sadly I think I'm in no position to explain `why` (an intended use case) but rather `how` (what the new behavior is); I just picked a `Good First Issue` ticket to force myself to dive in the source code :sweat_smile: . 

I think the original idea of use case was https://github.com/iterative/dvc/issues/4497 but I might misinterpreted the use case along the implementation.

Reviewer 2: Yeah, sorry for the lack of clarity here after you've already put in so much work @daavoo.

> but in this case we don't do anything to the .dvc or dvc.yaml?

I just tested, and it looks like .dvc is removed with this command but dvc.yaml won't be touched. Is that right @daavoo?

There might be enough straightforward utility here to merge the PR if it is limited to making `dvc remove data.xml` equivalent to `dvc remove data.xml.dvc` to support https://github.com/iterative/dvc/issues/2575#issuecomment-707868202.

The rest might be valuable, too, but it seems like the semantics need to be straightened out in https://github.com/iterative/dvc/issues/5791 first.

Author: If you consider that we need to take a step back and put this and the dvc P.R. on pause or even discarding them in favor of waiting to find the right semantics, I would not mind it at all. 

I wanted to dive into the source code and your thoughtful development process and I did; don't feel obligated for merging something just because of the work i already put.

usage: dvc remove [-h] [-q | -v] [--outs] targets [targets ...]

positional arguments:

>   targets        stages (found in dvc.yaml), .dvc files or name of output files to remove.

Reviewer 3: let's please not use the `output` term here . `output` is considered more of an internal DVC terminology, for end-users it's better to use data files, model files, directories - that kind of terminology. Usually,  something that anyone would immediately understand is better than a technical term

Reviewer 3: It feels the Description below, needs some update to it as well

especially part like this:

> `dvc remove` doesn't remove files from the DVC <abbr>cache</abbr> or
> [remote storage](/doc/command-reference/remote). Use `dvc gc` for that.

do we change this behavior? what is the most common scenario we now expect? etc

cc @dberenbaum

Author: > let's please not use the `output` term here . `output` is considered more of an internal DVC terminology, for end-users it's better to use data files, model files, directories - that kind of terminology. Usually, something that anyone would immediately understand is better than a technical term

@shcheklein I understand your point and I have been trying to find a better name, without much success. Technical writing in English is not my strength :sweat_smile: 

In this case, I feel that we need to use the `output` term here, in the same way it is being used in other places along the doc page, because we are actually talking about internal DVC terminology (i.e. we are referring to the `outpus` of a `stage`). 

For sure the semantics surrounding the `output` word can be improved and I will try to do that.

> It feels the Description below, needs some update to it as well
> 
> especially part like this:
> 
> > `dvc remove` doesn't remove files from the DVC cache or
> > [remote storage](/doc/command-reference/remote). Use `dvc gc` for that.
> 
> do we change this behavior? what is the most common scenario we now expect? etc
> 
> cc @dberenbaum

That behavior did not changed. I guess the confusion is my fault for not updating other parts of the doc page. The change introduced in the DVC P.R. changes the behavior described a little bellow that section:

```
Note that the actual output files or directories of the stage (outs field) are not removed by this command, unless the --outs option is used.
```

After the change in the DVC P.R. (and what should we properly document here), you can now remove *concrete* (i.e. one of many) output files or directories of the stage (outs field). Previously you could only remove *all* the outputs as described in that section.

Reviewer: > for end-users it's better to use data files, model files, directories - that kind of terminology

The problem with that may be that we specifically mean stage outputs, not dependencies (which can also be all of those things) — same as https://github.com/iterative/dvc/pull/5273#discussion_r609853710.

Reviewer: > we specifically mean stage outputs here, not dependencies

Actually I guess indeed "outputs" doesn't make that much sense in the context of add/remove.

> we are referring to the outpus of a stage

Not necessarily, @daavoo. .dvc files aren't stages and while they have the `outs` field inside that's a just the internal terminology like Ivan mentioned. How about:

```suggestion
  targets        Tracked files/directories, stage names (found in
                 dvc.yaml), or .dvc files to remove.
```

Reviewer: > The change introduced in the DVC P.R. changes the behavior described a little bellow that section:
> `Note that the actual output files or directories of the stage (outs field) are not removed...`

@daavoo how is that changed by iterative/dvc/pull/5757? In the case of stage name targets. Thanks

Author: My bad for not clarifying. There are no changes in the case of stage name targets.

I was referring to the last part of that statement `unless the --outs option is used.` and trying to say that `--outs` is no longer the only option to remove `the actual output files or directories of the stage`

Reviewer: Gotcha, sounds like we're on the same page now then. Thanks!

- `-v`, `--verbose` - displays detailed tracing information.

## Example: Remove stage outputs

Let's imagine we have a `train` stage in `dvc.yaml`, and corresponding files in
the <abbr>workspace</abbr>:

```yaml
train:
  cmd: python train.py data.csv
  deps:
    - data.csv
    - train.py
  outs:
    - logs
    - model.h5
```

```dvc
$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  model.h5  logs  train.py

$ cat .gitignore
/data.csv
/model.h5
/logs
```

Using `dvc remove` on the stage name will remove the stage from `dvc.yaml`, and
corresponding entries from `.gitignore`. With the `--outs` option, the actual
files and directories are deleted too (`logs/` and `model.h5` in this example):

```dvc
$ dvc remove train --outs

$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  train.py

$ cat .gitignore
/data.csv
```

> Notice that the dependencies (`data.csv` and `train.py`) are not deleted.

> ## Example: remove a specific stage output

Reviewer: Not sure we need this example AND the next one ("remove specific data" from .dvc files)

Assuming we have the same initial <abbr>workspace</abbr> as before:

```yaml
train:
  cmd: python train.py data.csv
  deps:
    - data.csv
    - train.py
  outs:
    - logs
    - model.h5
```

```dvc
$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  model.h5  logs  train.py

$ cat .gitignore
/data.csv
/model.h5
/logs
```

`dvc remove` can also be used on **individual** <abbr>output</abbr> files or
directories of a stage:

```dvc
$ dvc remove model.h5

$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  logs  train.py

$ cat .gitignore
/data.csv
/logs
```

> The <abbr>output</abbr> file is actually being removed from the

Reviewer 3: @daavoo besides removing it from the `.gitignore`, does it mean that we have pretty much `rm -f`?

Reviewer: Apparently. Is that a bug @shcheklein ? It feels like it should be smart enough to edit dvc.yaml in this case.

> The <abbr>output</abbr> file is actually being removed from the
> <abbr>workspace</abbr> and `.gitignore` but `dvc.yaml` is not being updated.

Reviewer: ```suggestion
`model.h5` file is removed from the <abbr>workspace</abbr> and `.gitignore`,
but note that `dvc.yaml` is not updated.
```

```yaml
train:

>   cmd: python train.py data.csv

Reviewer 3: thanks 👍

```dvc
$ dvc remove train --outs
$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  train.py
```

> Notice that the dependencies (`data.csv` and `train.py`) are not deleted.

> ## Example: remove a single output file by name
> 
> Imagine the same <abbr>workspace</abbr> as before but now we have multiple
> `outs` for the `train` stage:

Reviewer: ```suggestion
## Example: remove a stage output by filename

Imagine the we have multiple <abbr>outputs</abbr> for a `train` stage:
```

Reviewer: But why focus the example on stage outputs? What about making it about tracked data in general? Showing both removing an added dataset (dir), as well as a stage output (a model), for example. And/or update a previous example to mention the alternative to targeting .dvc files.

Author: You are totally right, I was too focused on stage outputs because that was the "use case" described in the issue. 

And now that I think about it, does it make sense to document targeting `.dvc` files *with* and *without*  the `.dvc` suffix?. I.e. why would a user prefer `dvc remove foo.csv.dvc` instead of just `dvc remove foo.csv`?

Reviewer: Happens. And we can go with this for now (just needs some fixes as indicated by Ivan) and I'll change it later. Unless you don't mind reworking it in a more general form along the way 🙂

Author: Documenting this change is being definitely more challenging than implementing it xD. 

@jorgeorpinel After reading https://github.com/iterative/dvc/issues/2575#issuecomment-707868202 , I'm thinking that maybe we should remove the example `dvc remove data.csv.dvc` in favor of `dvc remove data.csv`, given that the behavior is now the same but new users might don't be aware of the concept of `.dvc` files?

Author: The existing section `## Example: remove a tracked file` kind of overlaps in content with a the separate page `# How to Stop Tracking Data`.

Reviewer: Thanks for the updates @daavoo, I can wrap this up once https://github.com/iterative/dvc/pull/5757 is merged.

```yaml
train:
  cmd: python train.py data.csv
  deps:
    - data.csv
    - train.py
  outs:
    - logs
    - model.h5
```

```dvc
$ ls
dvc.lock  dvc.yaml  data.csv  data.csv.dvc  logs  model.h5  train.py
```

> Using `dvc remove` you can remove a specific file using it's output name:

Reviewer 3: what is the semantics of this operation? it is not clear from this update, unfortunately - do we remove outputs from the workspace? from the cache? do we update the `dvc.yaml` and/or remove .dvc along the way?

Author: I have tried to update the semantics for better explaining the new behavior.
